src/main/scala/org/globalforestwatch/features/FeatureRDD.scala

```diff
@@ -60,92 +71,137 @@ object FeatureRDD {
+          )
       }
 
-    // In case we implement this method for other geometry types we will have to split geometries
-    //      .flatMap { feature =>
-    //        splitGeometry(feature)
-    //      }
-
   }
 
-
   /*
     Convert polygon-polygon intersection join to feature RDD
    */
   def apply(
-             feature1Obj: Feature,
-             feature2Obj: Feature,
-             spatialRDD: SpatialRDD[GeoSparkGeometry],
+             feature1Type: String,
+             feature1Uris: NonEmptyList[String],
+             feature1Delimiter: String,
+             feature2Type: String,
+             feature2Uris: NonEmptyList[String],
+             feature2Delimiter: String,
              kwargs: Map[String, Any],
+             spark: SparkSession
            ): RDD[geotrellis.vector.Feature[Geometry, FeatureId]] = {
-    val scalaRDD =
-      org.apache.spark.api.java.JavaRDD.toRDD(spatialRDD.spatialPartitionedRDD)
-
-    scalaRDD
-      .flatMap {
-        case shp: GeoSparkPolygonal =>
-          val featureData = shp.getUserData.asInstanceOf[String].split('\t')
-          val writer: WKTWriter = new WKTWriter()
-          val wkt = writer.write(shp)
-
-          val reader: WKTReader = new WKTReader()
-          val geom = GeometryReducer.reduce(GeometryReducer.gpr)(
-            reader.read(wkt)
-          )
 
-          val feature1Data = featureData.head.drop(1).dropRight(1).split(',')
-          val feature1Id: FeatureId =
-            feature1Obj.getFeatureId(feature1Data, parsed = true)
-
-          val feature2Data = featureData.tail.head.drop(1).dropRight(1).split(',')
-          val feature2Id: FeatureId =
-            feature2Obj.getFeatureId(feature2Data, parsed = true)
+    val spatialDF: DataFrame = PolygonIntersectionDF(
+      feature1Uris,
+      feature1Type,
+      feature2Uris,
+      feature2Type,
+      spark,
+      kwargs,
+      feature2Delimiter = ","
+    )
+
+    val pairedRDD = spatialDF.rdd.map { row: Row =>
+      val featureId1: FeatureId = FeatureIdFactory(feature1Type)
+        .fromUserData(row.getAs[Row](0).toString, ",")
+      val featureId2: FeatureId = FeatureIdFactory(feature2Type)
+        .fromUserData(row.getAs[Row](1).toString, ",")
+      val geom = row.getAs[GeoSparkGeometry](2)
+
+      (CombinedFeatureId(featureId1, featureId2), geom)
+    }
 
-          Some(vector.Feature(geom, CombinedFeatureId(feature1Id, feature2Id)))
-        case _ =>
-          // Polygon-polygon intersections can generate points or lines, which we just want to ignore
-          logger.warn(
-            "Cannot process geometry type"
-          )
-          None
+    val hashPartitioner = new HashPartitioner(pairedRDD.getNumPartitions)
+    pairedRDD
+      .keyBy({ pair: (CombinedFeatureId, GeoSparkGeometry) =>
+        Z2(
+          (pair._2.getCentroid.getX * 100).toInt,
+          (pair._2.getCentroid.getY * 100).toInt
+        ).z
+      })
+      .partitionBy(hashPartitioner)
```

Comment:
To make sure data are not skewed I apply a hash partitioner here. Later I use the regular range partitioner for burned areas again.